Wikipedia/Code/ArticleFetcher.swift

```diff
@@ -281,6 +281,79 @@ final public class ArticleFetcher: Fetcher, CacheFetching {
         }
     }
     
+    /// Makes periodic HEAD requests to the mobile-html endpoint until the etag no longer matches the one provided.
+    @discardableResult public func waitForMobileHTMLChange(articleURL: URL, eTag: String, timeout: TimeInterval, cancellationKey: CancellationKey? = nil, completion: @escaping (Result<String, Error>) -> Void) -> CancellationKey? {
+        guard timeout > 0 else {
+            completion(.failure(RequestError.timeout))
+            return nil
+        }
+        let requestURL: URL
+        do {
+            requestURL = try mobileHTMLURL(articleURL: articleURL)
+        } catch let error {
+            completion(.failure(error))
+            return nil
+        }
+        let start = CFAbsoluteTimeGetCurrent()
```

Comment:
Just out of curiosity: is the intention behind using Foundation's time utility here over something higher level like `Date` just for performance characteristic improvements? I.e. avoiding a slow heavyweight `Date`. If I'm not mistaken, this layer of time utility is still affected by the system clock, same as `Date` is. So there might be funny business if the user manipulates the clock during the initial 7 second timeout, or the clock is changed under them by the system/DST/NTP. From what I understand `mach_absolute_time` provides a system clock agnostic way to performantly and precisely handle duration. Though again to be clear, I'm not suggesting we change this or use that – in practice, this seems fine. Just curious.

Reply:
Infinite loop potential is a bit scary here, we could also alternatively (or in addition to this) try a `maximumIterationCount` and force them out if they've reached beyond a certain number there.

Reply:
> Just out of curiosity: is the intention behind using Foundation's time utility here over something higher level like `Date` just for performance characteristic improvements? I.e. avoiding a slow heavyweight `Date`. If I'm not mistaken, this layer of time utility is still affected by the system clock, same as `Date` is. So there might be funny business if the user manipulates the clock during the initial 7 second timeout, or the clock is changed under them by the system/DST/NTP. From what I understand `mach_absolute_time` provides a system clock agnostic way to performantly and precisely handle duration. Though again to be clear, I'm not suggesting we change this or use that – in practice, this seems fine. Just curious.

No specific intention here, thanks for flagging this as the system time changing could cause an issue. I did some digging and it looks like `DispatchTime` is a good somewhat high-level system-time-zone-agnostic option. I've updated the PR to use `DispatchTime`

Reply:
> Infinite loop potential is a bit scary here, we could also alternatively (or in addition to this) try a `maximumIterationCount` and force them out if they've reached beyond a certain number there.

@tonisevener I very well could be missing something, but if `DispatchTime` is unaffected by the system clock (that is, its value _always_ increases in "tick" units), then we're guaranteed that `duration` is always non-negative and recursive calls into `waitForMobileHTMLChange` are always with timeout values _less than_ the originally seeded value (7 seconds). Given that, seems like the `guard timeout > 0` will always cover us and obviate the need for a `maximumIterationCount`, because it's essentially functionally equivalent to one. What do you think?

Reply:
After thinking about this some more, I updated this to use `maxAttempts` instead of `timeout`. It allowed for an easier way to make each attempt wait longer in-between instead of making a request every quarter second. Also, given that `Double`s can do strange things sometimes, even though in theory the timeout should be safe, I agree with Toni's sentiment and feel better about having the loop breaking condition being an integer check.

Reply:
Thanks, it's working well for me. @staykids agreed, I think I would have approved either approach - `DispatchTime` felt much safer than `CFAbsoluteTimeGetCurrent ` (great catch on that).